### ISSUE TITLE
Failed Testing & Weird Behavior

### ISSUE THREAD

**Issue by eyildiz-ugoe** [nerian-vision/nerian_stereo#8]
So the testing commands given in the node's page are not showing any image to me. The window pops up, but the content is fully black.

![screenshot from 2019-02-15 11-11-12](https://user-images.githubusercontent.com/35396058/52850259-56170680-3113-11e9-9659-5a6e3363772e.png)

Everything works just fine in the web interface, the images are shown, I also calibrated it and put the calibration file under /data, referred to it in the launch file, all here is OK. However, there is a problem with the data transfer or something, I don't know, I cannot get the image via launching the `image_view` commands. It's all dark.

Moreover, I launched Rviz and noticed that it showed the image_left and pointcloud2 once only, then got stuck and never updated it. image_right was not even there at all. Sometimes there is no image at all, like below:

![12](https://user-images.githubusercontent.com/35396058/52868726-61365a80-3144-11e9-81d2-de50327c96e0.png)

In short, ROS node does not seem to be functioning well.

EDIT: At first I installed the node using the guide, however that does not install it properly and you get an error "download.sh isn't there" and such, hence I cloned this repository and moved it to `/opt/ros/melodic/share/` where the node was previously installed. If this breaks anything or if this is wrong, let me know. And in that case, you need to edit the ROS node page since it is misleading (using that installation command isn't sufficient, since the contents are different from what you have here in this repo).

**Comment by nerian-vision** [comment]
The ROS node is not receiving any data. A likely explanation is that SceneScan's firmware version doesn't match. We significantly changed the network protocol with firmware version 2.0. Can you please update SceneScan to the latest version? Firmware downloads are on the Nerian website:
https://nerian.com/support/software/

**Comment by eyildiz-ugoe** [comment]
Hi,

I don't think this is the case, because the web interface shows that the version is 3.1.2, which is the latest.
Could you tell me how to install the ROS package manually? The one that comes with sudo apt-get is not working properly, and it has even missing files such as the .sh scripts, so it gives error when the node is launched. (This version should not be suggested to users in the tutorial, in ROS melodic's case, the package downloaded has missing files.).
I cloned the one from the Github repository, what to do after this step? (This also needs to be on the tutorial page at some point).
Regards,

--
-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~
Erenus Yildiz
PhD Student

Georg-August-Universität Göttingen
III. Physikalisches Institut - Biophysik
Friedrich-Hund-Platz 1
37077 Göttingen

Phone: +49 (0) 176 726 542 51
Email: eyildiz@gwdg.de (https://link.getmailspring.com/link/1550243972.local-1e313e3f-a168-v1.5.6-4cb1851b@getmailspring.com/0?redirect=mailto%3Aeyildiz%40gwdg.de&recipient=cmVwbHkrMDIxYzE5ZGExMzBjMzY1NTU4NzU4OGZhYjM5NGRmNGY0ODkzMWMzY2JlNjg0NzlhOTJjZjAwMDAwMDAxMTg3ZTg4MDc5MmExNjljZTE4N2FiZGRjQHJlcGx5LmdpdGh1Yi5jb20%3D)
-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~-~

**Comment by nerian-vision** [comment]
I see that in the current release these scripts didn't have the right permissions. Which is why you got the error. This will be fixed.

If you build the package yourself then you need to do this with catkin:
http://wiki.ros.org/catkin/Tutorials

For finding the cause of the connection problem, I would recommend you to try connecting with our NVCom application first. You can download the sources and build instructions form the Nerian website.

**Comment by eyildiz-ugoe** [comment]
Now that I noticed, the scripts under `/build/libvisiontransfer/examples` don't work either. Each of them gives `No devices discovered!` error, while I can connect to the web interface with no problem. Something is definitely wrong.

I did what you suggested and launched `nvcom`, no surprise, that also said there was no device, while everything on the web interface was OK.

![1](https://user-images.githubusercontent.com/35396058/52868627-2af8db00-3144-11e9-9f93-84a92625f6ff.png)

**Comment by nerian-vision** [comment]
The web interface uses a different protocol. It seems that you don't receive broadcast messages from SceneScan. Is SceneScan connected directly or through a switch or router? Does your firewall permit broadcast messages?

**Comment by eyildiz-ugoe** [comment]
There is no router/switch inbetween, no. However, today I made some tests and came to interesting conclusions.

I actually got some outputs, but whatever I tried, I never could receive any data from the topic `/right_image`, and I tried this on 3 different computers.  I am  able to get a fluent data flow from the pointcloud and left_image, but not the right one. See below. In the upper-right terminal, I used `rostopic echo`on the `right_image` and there is no output. The data above was from the `left_image` which works fine.

![1](https://user-images.githubusercontent.com/35396058/52903744-4803db80-3222-11e9-93e7-5254d7844328.png)

This is really strange, and I can't think of any reason. I can't really blame the cables we use, like if the sent data is not transferred or something, because the standalone web interface works just fine, I can see the right camera's image on the calibration page. But there is no ros publication when I run that command as you see above.

I am really puzzled by all this, what do you make out of this?

**Comment by eyildiz-ugoe** [comment]
Even more interesting observations:

When I run nvcom, this is the result I see:

![ss1](https://user-images.githubusercontent.com/35396058/52904035-548a3300-3226-11e9-9831-0387aad3fab2.png)


And when I run some examples (such as OpenCV one), I see the following images being saved. What's up with the right image? 

![image0](https://user-images.githubusercontent.com/35396058/52904040-5eac3180-3226-11e9-9643-e995b2d00387.png)
![image1](https://user-images.githubusercontent.com/35396058/52904041-5eac3180-3226-11e9-8230-075c12e469cb.png)

**Comment by nerian-vision** [comment]
The right image is usually not transmitted over the network. The right image is only output if SceneScan is set to pass-through or rectify mode. This is done in order tor preserve network bandwidth.

The reason why the disparity map that is written from the OpenCV image appears so dark is that this is a 16-bit image file. Our disparity maps are, however, only 12-bits. This is why the pixels appear very dark. Unfortunately OpenCV doesn't support any 12-bit formats. If you use ImagePair::writePgmFile() from our API, this will correctly write a 12-bit PGM file.

For viewing the 16-bit PNG files, I recommend using ImageJ, which can correctly deal with us using only a subset of the available 16-bit value range.

**Comment by eyildiz-ugoe** [comment]
The problem is not about OpenCV, ROS does not publish any data for `right_image` either. Something is wrong with the transmission of right image.

**Comment by nerian-vision** [comment]
The right image will be published if you change the operation mode to pass-through or rectify (see above).

The image written by the OpenCV example is correct, it's just difficult to visualize as our value range is from 0 to 4095, and a 16-bit image has a value range from 0 to 65535. For disparities the absolute magnitude is relevant which is why we can't scale these values without scaling them back when doing any onward processing.

**Comment by eyildiz-ugoe** [comment]
Is there any possibility of having images in RGB color space? So far I only attained grayscaled images from the left camera.

**Comment by nerian-vision** [comment]
If you have an RGB camera then you can get RGB images
